src/lib.rs

```diff
@@ -522,6 +569,54 @@ pub mod pallet {
 		pub fn pot() -> BalanceOf<T> {
 			T::RewardCurrency::free_balance(&Self::account_id())
 		}
+		/// Verify a set of signatures made with relay chain accounts
+		/// We are verifying all the signatures, and then counting
+		/// We could do something more efficient like count as we verify
+		/// In any of the cases the weight will need to account for all the signatures,
+		/// as we dont know beforehand whether they will be valid
+		fn verify_signatures(
+			proofs: Vec<(T::RelayChainAccountId, MultiSignature)>,
+			reward_info: RewardInfo<T>,
+			payload: Vec<u8>,
+		) -> DispatchResult {
+			// The proofs should
+			// 1. be signed by contributors to this address, otherwise they are not counted
+			// 2. Signs a valid native identity
+			// 3. The sum of the valid proofs needs to be bigger than UnsifficientNumberOfValidProofs
+
+			// I use a map here for faster lookups
+			let mut voted: BTreeMap<T::RelayChainAccountId, ()> = BTreeMap::new();
+			for (relay_account, signature) in proofs {
+				// We just count votes that we have not seen
+				if voted.get(&relay_account).is_none() {
```

Comment:
What if I made ten contributions from the same relay account? Is the stored list of contributing accounts guaranteed not to have duplicates? If the stored list has that same address ten times, and you only count it once here, it will be impossible to change the rewards account.

Reply:
I think https://github.com/PureStake/crowdloan-rewards/blob/0d7f6fdbdc5ae2de3e18ddf200188d7c810de963/src/lib.rs#L429 prevents this. At initialization level, we guarantee that you dont initialize twice the same contribution address.

Reply:
The case you are describing refers to ten contributions from the same relay account to the same reward address (as the contributor info is stored in a per-address mapping), but it's not possible to initialize the crowdloan-rewards pallet like that